### ISSUE TITLE
Unknown instance error and template not found error appear in the mixer logs

### ISSUE THREAD

**Issue by vadimeisenbergibm** [istio/istio#13490]
Perform the following steps:

1.  Create a `quota` instance:

    ```
    kubectl apply -f - <<EOF
    apiVersion: config.istio.io/v1alpha2
    kind: instance
    metadata:
      name: requestcountquota
    spec:
      compiledTemplate: quota
      params:
        dimensions:
          destination: destination.labels["app"] | destination.service.name | "unknown"
    EOF
    ```

Mixer log:

```
2019-04-20T08:28:40.719973Z	info	Publishing 2 events
2019-04-20T08:28:40.723360Z	error	template '' not found
2019-04-20T08:28:40.723627Z	info	Built new config.Snapshot: id='26'
2019-04-20T08:28:40.724014Z	info	adapters	getting kubeconfig from: ""	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:28:40.724028Z	warn	Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2019-04-20T08:28:40.725009Z	info	adapters	Waiting for kubernetes cache sync...	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:28:41.271796Z	info	adapters	Cache sync successful.	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:28:41.271851Z	info	adapters	getting kubeconfig from: ""	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:28:41.271863Z	warn	Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2019-04-20T08:28:41.278205Z	info	Setting up event handlers
2019-04-20T08:28:41.278342Z	info	Starting Secrets controller
2019-04-20T08:28:41.278463Z	info	Waiting for informer caches to sync
gc 135 @13689.904s 0%: 0.52+401+96 ms clock, 2.1+198/418/195+387 ms cpu, 18->20->12 MB, 19 MB goal, 6 P
2019-04-20T08:28:41.470943Z	info	Cleaning up handler table, with config ID:25
2019-04-20T08:28:41.470978Z	info	adapters	deleted remote controller 	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:28:41.471010Z	info	adapters	adapter closed all scheduled daemons and workers	{"adapter": "stdio.istio-system"}
2019-04-20T08:28:41.471145Z	info	adapters	shutting down daemon...	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:28:41.471150Z	info	adapters	shutting down daemon...	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:28:41.471169Z	info	adapters	shutting down daemon...	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:28:41.471206Z	info	adapters	adapter closed all scheduled daemons and workers	{"adapter": "prometheus.istio-system"}
2019-04-20T08:28:41.471261Z	info	adapters	shutting down daemon...	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:28:42.471658Z	info	adapters	adapter closed all scheduled daemons and workers	{"adapter": "kubernetesenv.istio-system"}
scvg91: 0 MB released
scvg91: inuse: 23, idle: 5, sys: 28, released: 1, consumed: 26 (MB)
```

1.  Create a handler:

    ```
    kubectl apply -f - <<EOF
    apiVersion: config.istio.io/v1alpha2
    kind: handler
    metadata:
      name: quotahandler
    spec:
      compiledAdapter: memquota
      params:
        quotas:
        - name: requestcountquota
          maxAmount: 500
          validDuration: 1s
    EOF
    ```

```
2019-04-20T08:32:38.324055Z	info	Publishing 4 events
2019-04-20T08:32:38.327625Z	error	template '' not found
2019-04-20T08:32:38.327765Z	info	Built new config.Snapshot: id='27'
2019-04-20T08:32:38.327994Z	info	adapters	getting kubeconfig from: ""	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:32:38.328009Z	warn	Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2019-04-20T08:32:38.328986Z	info	adapters	Waiting for kubernetes cache sync...	{"adapter": "kubernetesenv.istio-system"}
gc 137 @13927.309s 0%: 1.3+395+1.7 ms clock, 5.4+381/608/0+6.9 ms cpu, 18->21->12 MB, 19 MB goal, 6 P
2019-04-20T08:32:39.029179Z	info	adapters	Cache sync successful.	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:32:39.029230Z	info	adapters	getting kubeconfig from: ""	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:32:39.029237Z	warn	Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2019-04-20T08:32:39.030247Z	info	Setting up event handlers
2019-04-20T08:32:39.030354Z	info	Starting Secrets controller
2019-04-20T08:32:39.030469Z	info	Waiting for informer caches to sync
2019-04-20T08:32:39.033548Z	info	Cleaning up handler table, with config ID:26
2019-04-20T08:32:39.033607Z	info	adapters	deleted remote controller 	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:32:39.033689Z	info	adapters	adapter closed all scheduled daemons and workers	{"adapter": "prometheus.istio-system"}
2019-04-20T08:32:39.033707Z	info	adapters	adapter closed all scheduled daemons and workers	{"adapter": "stdio.istio-system"}
2019-04-20T08:32:39.033975Z	info	adapters	shutting down daemon...	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:32:39.034013Z	info	adapters	shutting down daemon...	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:32:39.034029Z	info	adapters	shutting down daemon...	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:32:39.034042Z	info	adapters	shutting down daemon...	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:32:40.033817Z	info	adapters	adapter closed all scheduled daemons and workers	{"adapter": "kubernetesenv.istio-system"}
```

1.  Create a rule:

    ```
    kubectl apply -f - <<EOF
    apiVersion: config.istio.io/v1alpha2
    kind: rule
    metadata:
      name: quota
    spec:
      actions:
      - handler: quotahandler
        instances:
        - requestcountquota
    EOF
    ```

```
2019-04-20T08:36:10.014908Z	info	Publishing 9 events
2019-04-20T08:36:10.018529Z	error	template '' not found
2019-04-20T08:36:10.018747Z	error	Instance not found: instance='requestcountquota'
2019-04-20T08:36:10.018769Z	error	No valid instances found
2019-04-20T08:36:10.018775Z	error	No valid actions found in rule
2019-04-20T08:36:10.018781Z	info	Built new config.Snapshot: id='28'
2019-04-20T08:36:10.019058Z	info	adapters	getting kubeconfig from: ""	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:36:10.019084Z	warn	Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2019-04-20T08:36:10.020090Z	info	adapters	Waiting for kubernetes cache sync...	{"adapter": "kubernetesenv.istio-system"}
gc 139 @14139.110s 0%: 0.016+394+1.0 ms clock, 0.067+392/713/202+4.2 ms cpu, 20->20->11 MB, 21 MB goal, 6 P
2019-04-20T08:36:10.720276Z	info	adapters	Cache sync successful.	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:36:10.720401Z	info	adapters	getting kubeconfig from: ""	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:36:10.720436Z	warn	Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2019-04-20T08:36:10.721655Z	info	Setting up event handlers
2019-04-20T08:36:10.721807Z	info	Starting Secrets controller
2019-04-20T08:36:10.721882Z	info	Waiting for informer caches to sync
2019-04-20T08:36:10.770372Z	info	Cleaning up handler table, with config ID:27
2019-04-20T08:36:10.770536Z	info	adapters	deleted remote controller 	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:36:10.770572Z	info	adapters	adapter closed all scheduled daemons and workers	{"adapter": "stdio.istio-system"}
2019-04-20T08:36:10.770605Z	info	adapters	adapter closed all scheduled daemons and workers	{"adapter": "prometheus.istio-system"}
2019-04-20T08:36:10.770796Z	info	adapters	shutting down daemon...	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:36:10.770823Z	info	adapters	shutting down daemon...	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:36:10.770847Z	info	adapters	shutting down daemon...	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:36:10.770864Z	info	adapters	shutting down daemon...	{"adapter": "kubernetesenv.istio-system"}
2019-04-20T08:36:10.770881Z	info	adapters	adapter closed all scheduled daemons and workers	{"adapter": "kubernetesenv.istio-system"}
```

Notice the errors:
```
2019-04-20T08:36:10.018529Z	error	template '' not found
2019-04-20T08:36:10.018747Z	error	Instance not found: instance='requestcountquota'
2019-04-20T08:36:10.018769Z	error	No valid instances found
```

`template '' not found` appears for all the items.

The quota, the handler and the rule are defined in the default namespace:

```
kubectl get instance,handler,rule -n default
NAME                                         AGE
instance.config.istio.io/requestcountquota   16m

NAME                                   AGE
handler.config.istio.io/quotahandler   12m

NAME                         AGE
rule.config.istio.io/quota   8m
```

Cleanup:

```
kubectl delete rule quota
kubectl delete handler quotahandler
kubectl delete instance requestcountquota
```

Istio version 1.1.2, deployed with istio-values-auth.yaml

**Comment by douglas-reid** [comment]
What happens if you apply these configs in the `istio-system` namespace?

**Comment by vadimeisenbergibm** [comment]
@douglas-reid I will try, but my guess is that these should work with a regular namespace too. The common use case with Istio, in my understanding, is that there are multiple teams, each working in their own namespace. These teams can create quotas for their namespace and put the configuration artifacts in their own namespace. The teams should not have write access to the common `istio-system` namespace.

**Comment by kyessenov** [comment]
The issue is the version of istio. Please try with istio 1.1.3 if you use "compiledTemplate" field. This was a bug fix that only made it in 1.1.3. There should be now a sample that does not use compiledTemplate, e.g. https://raw.githubusercontent.com/istio/istio/release-1.1/samples/bookinfo/policy/mixer-rule-productpage-ratelimit-crd.yaml

**Comment by vadimeisenbergibm** [comment]
@kyessenov Thanks, these errors disappeared in 1.1.3. Closing the issue.
